update Isabelle 2021

lsf37 opened the pull request:
This PR updates the repo to work with Isabelle2021, re-syncing AFP sessions with AFP2021.

This needs either the `ts-2021` branch or `Isabelle2021` tag in `isabelle`, which the PR tests won't have yet until the manifest is updated, so we expect these tests to fail. Will run a separate testboard for checking.

Testboard here: https://github.com/seL4/gh-testboard/runs/3750528531

lsf37 commented on lib/Lib.thy:
```diff
@@ -19,10 +19,11 @@ imports
+  "Word_Lib.WordSetup"
 begin
 
+typ "32 word"
+
```
- [ ] remove stray `typ`

lsf37 commented on proof/crefine/ARM/Refine_C.thy:
```diff
@@ -241,6 +230,10 @@ lemma ct_active_not_idle'_strengthen:
   "invs' s \<and> ct_active' s \<longrightarrow> ksCurThread s \<noteq> ksIdleThread s"
   by clarsimp
 
+lemma uint_minus_1_eq:
+  \<open>uint (- 1 :: 'a word) = 2 ^ LENGTH('a::len) - 1\<close>
+  by transfer (simp add: take_bit_minus_one_eq_mask mask_eq_exp_minus_1)
+
```
- [ ] move to Word_Lib

lsf37 commented on proof/crefine/ARM/Tcb_C.thy:
```diff
@@ -3153,6 +3156,10 @@ lemma not_isThreadCap_case:
    (case cap of ThreadCap x \<Rightarrow> f x | _ \<Rightarrow> g) = g"
   by (clarsimp simp: isThreadCap_def split: capability.splits)
 
+lemma FF_eq_minus_1:
+  \<open>0xFF = (- 1 :: 8 word)\<close>
+  by simp
```
- [ ] move to Word_Lib internal

lsf37 commented on proof/crefine/ARM_HYP/Tcb_C.thy:
```diff
@@ -3246,6 +3249,10 @@ lemma not_isThreadCap_case:
    (case cap of ThreadCap x \<Rightarrow> f x | _ \<Rightarrow> g) = g"
   by (clarsimp simp: isThreadCap_def split: capability.splits)
 
+lemma FF_eq_minus_1:
+  \<open>0xFF = (- 1 :: 8 word)\<close>
+  by simp
+
```
Remove (copy should be in Word_lib)

lsf37 commented on proof/crefine/RISCV64/Recycle_C.thy:
```diff
@@ -260,7 +260,7 @@ lemma unat_of_nat_pageBitsForSize_32 [simp]:
   done
 
-(* FIXME move *)
+(* FIXME move to Word_Lib *)
 lemma shiftl_t2n':
   "w << n = w * (2 ^ n)"
   for w :: "'a::len word"
```
should probably perform that move

lsf37 commented on proof/crefine/RISCV64/Refine_C.thy:
```diff
@@ -232,6 +231,10 @@ lemma ct_active_not_idle'_strengthen:
 
 
 
+lemma uint_minus_1_eq:
+  \<open>uint (- 1 :: 'a word) = 2 ^ LENGTH('a::len) - 1\<close>
+  by transfer (simp add: take_bit_minus_one_eq_mask mask_eq_exp_minus_1)
+
```
remove (copy should exist)

lsf37 commented on proof/crefine/RISCV64/Tcb_C.thy:
```diff
@@ -3203,6 +3206,10 @@ lemma not_isThreadCap_case:
    (case cap of ThreadCap x \<Rightarrow> f x | _ \<Rightarrow> g) = g"
   by (clarsimp simp: isThreadCap_def split: capability.splits)
 
+lemma FF_eq_minus_1:
+  \<open>0xFF = (- 1 :: 8 word)\<close>
+  by simp
+
```
remove

lsf37 commented on proof/crefine/X64/Refine_C.thy:
```diff
@@ -232,6 +231,10 @@ lemma ct_active_not_idle'_strengthen:
 
 
 
+lemma uint_minus_1_eq:
+  \<open>uint (- 1 :: 'a word) = 2 ^ LENGTH('a::len) - 1\<close>
+  by transfer (simp add: take_bit_minus_one_eq_mask mask_eq_exp_minus_1)
+
```
remove

lsf37 commented on proof/crefine/X64/Tcb_C.thy:
```diff
@@ -3196,6 +3199,10 @@ lemma not_isThreadCap_case:
    (case cap of ThreadCap x \<Rightarrow> f x | _ \<Rightarrow> g) = g"
   by (clarsimp simp: isThreadCap_def split: capability.splits)
 
+lemma FF_eq_minus_1:
+  \<open>0xFF = (- 1 :: 8 word)\<close>
+  by simp
+
```
remove

lsf37 commented on proof/invariant-abstract/Detype_AI.thy:
```diff
@@ -929,7 +927,7 @@ lemma of_nat_le_pow:
   done
 
-lemma maxword_len_conv': "(x::machine_word) + max_word = x - 1" by (simp add: max_word_def)
+lemma maxword_len_conv': True by simp
```
remove

lsf37 commented on tools/autocorres/tests/examples/FactorialTest.thy:
```diff
@@ -52,6 +52,9 @@ proof (induct n arbitrary: m rule: less_induct)
     done
 qed
 
+(* FIXME: set default cases and induct rules for word *)
+declare word_induct2[induct type]
+
```
do that!

lsf37 commented on tools/autocorres/tests/examples/WordAbs.thy:
```diff
@@ -211,8 +211,8 @@ lemma "n \<ge> 32 \<Longrightarrow> \<not> no_fail \<top> (U_shiftl_S_abs_S' (x
          U_shiftl_S_abs_S' (x::word32) (n::int)
-       \<lbrace>\<lambda>r s. r = x << nat n\<rbrace>!"
-  by (wpsimp simp: U_shiftl_S_abs_S'_def UINT_MAX_def unat_of_int)
+       \<lbrace>\<lambda>r s. r = x << nat n\<rbrace>!" thm U_shiftl_S_abs_S'_def
```
remove `thm`

lsf37 commented on tools/autocorres/tests/examples/WordAbs.thy:
```diff
@@ -229,11 +229,11 @@ lemma "x < 0 \<Longrightarrow> \<not> no_fail \<top> (S_shiftl_U_abs_US' (x :: i
+lemma "\<lbrace>\<lambda>s. n < 32 \<and> 0 \<le> x \<and> x << n \<le> INT_MAX\<rbrace>
          S_shiftl_U_abs_US' (x::int) (n::nat)
-       \<lbrace>\<lambda>r s. r = x << nat n\<rbrace>!"
+       \<lbrace>\<lambda>r s. r = x << n\<rbrace>!" thm S_shiftl_U_abs_US'_def
```
remove `thm`

lsf37 commented:
This is not exactly a tiny update, with over 480 files changed, but most files only have a few lines change, so it's not that terrible.

I would suggest that each reviewer pick a few sessions, commenting first which one they want to claim, and then go through those.

In terms of content, the main reason for change is a re-design of the word library in HOL and Word_Lib in AFP by Florian Haftmann. Most operations are now defined on abstract algebraic structures, and words are just one of multiple instances (nat and int are others for instance). Overall things seem to be getting simpler. The main thing to look out for is the occurrence of `take_bit n x`, mostly after automatic conversion to `nat` or `int`. It's somewhat similar to the previous `bintrunc` for those who remember that.  The bundle `no_take_bit` removes the main rules that produce such terms.

mbrcknl commented on proof/invariant-abstract/ARM/ArchIpc_AI.thy:
```diff
@@ -422,60 +449,15 @@ lemma transfer_caps_loop_valid_vspace_objs[wp, Ipc_AI_assms]:
         | assumption | simp split del: if_split)+
   done
 
-crunch aligned                   [wp, Ipc_AI_assms]:  make_arch_fault_msg "pspace_aligned"
-crunch distinct                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "pspace_distinct"
-crunch vmdb                      [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_mdb"
-crunch vmdb                      [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_mdb"
-crunch ifunsafe                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "if_unsafe_then_cap"
-crunch iflive                    [wp, Ipc_AI_assms]:  make_arch_fault_msg "if_live_then_nonz_cap"
-crunch state_refs_of             [wp, Ipc_AI_assms]:  make_arch_fault_msg "\<lambda>s. P (state_refs_of s)"
-crunch ct                        [wp, Ipc_AI_assms]:  make_arch_fault_msg "cur_tcb"
-crunch zombies                   [wp, Ipc_AI_assms]:  make_arch_fault_msg "zombies_final"
-crunch it                        [wp, Ipc_AI_assms]:  make_arch_fault_msg "\<lambda>s. P (idle_thread s)"
-crunch valid_globals             [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_global_refs"
-crunch reply_masters             [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_reply_masters"
-crunch valid_idle                [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_idle"
-crunch arch                      [wp, Ipc_AI_assms]:  make_arch_fault_msg "\<lambda>s. P (arch_state s)"
-crunch typ_at                    [wp, Ipc_AI_assms]:  make_arch_fault_msg "\<lambda>s. P (typ_at T p s)"
-crunch irq_node                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "\<lambda>s. P (interrupt_irq_node s)"
-crunch valid_reply               [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_reply_caps"
-crunch irq_handlers              [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_irq_handlers"
-crunch vspace_objs               [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_vspace_objs"
-crunch global_objs               [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_global_objs"
-crunch global_vspace_mapping     [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_global_vspace_mappings"
-crunch arch_caps                 [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_arch_caps"
-crunch v_ker_map                 [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_kernel_mappings"
-crunch eq_ker_map                [wp, Ipc_AI_assms]:  make_arch_fault_msg "equal_kernel_mappings"
-crunch asid_map                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_asid_map"
-crunch only_idle                 [wp, Ipc_AI_assms]:  make_arch_fault_msg "only_idle"
-crunch pspace_in_kernel_window   [wp, Ipc_AI_assms]:  make_arch_fault_msg "pspace_in_kernel_window"
-crunch cap_refs_in_kernel_window [wp, Ipc_AI_assms]:  make_arch_fault_msg "cap_refs_in_kernel_window"
-crunch valid_objs                [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_objs"
-crunch valid_ioc                 [wp, Ipc_AI_assms]:  make_arch_fault_msg "valid_ioc"
-crunch pred_tcb                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "pred_tcb_at proj P t"
-crunch cap_to                    [wp, Ipc_AI_assms]:  make_arch_fault_msg "ex_nonz_cap_to p"
-crunch pred_tcb                  [wp, Ipc_AI_assms]:  make_arch_fault_msg "pred_tcb_at proj P t"
```
Were these already redundant, or has something changed to make this possible?

lsf37 replied:
They probably became redundant a while ago, but most of the crunch removals only worked on ARM

mbrcknl commented on proof/invariant-abstract/ARM_HYP/ArchIpc_AI.thy:
```diff
@@ -13,10 +13,36 @@ context Arch begin global_naming ARM_HYP
-
-crunch cap_refs_respects_device_region[wp]: set_extra_badge "cap_refs_respects_device_region"
-  (wp: crunch_wps cap_refs_respects_device_region_dmo)
+lemma update_cap_data_closedform:
+  "update_cap_data pres w cap =
+   (case cap of
+     cap.EndpointCap r badge rights \<Rightarrow>
+       if badge = 0 \<and> \<not> pres then (cap.EndpointCap r (w && mask 28) rights) else cap.NullCap
+   | cap.NotificationCap r badge rights \<Rightarrow>
+       if badge = 0 \<and> \<not> pres then (cap.NotificationCap r (w && mask 28) rights) else cap.NullCap
+   | cap.CNodeCap r bits guard \<Rightarrow>
+       if word_bits < unat ((w >> 3) && mask 5) + bits
+       then cap.NullCap
+       else cap.CNodeCap r bits ((\<lambda>g''. drop (size g'' - unat ((w >> 3) && mask 5)) (to_bl g'')) ((w >> 8) && mask 18))
+   | cap.ThreadCap r \<Rightarrow> cap.ThreadCap r
+   | cap.DomainCap \<Rightarrow> cap.DomainCap
+   | cap.UntypedCap d p n idx \<Rightarrow> cap.UntypedCap d p n idx
+   | cap.NullCap \<Rightarrow> cap.NullCap
+   | cap.ReplyCap t m rights \<Rightarrow> cap.ReplyCap t m rights
+   | cap.IRQControlCap \<Rightarrow> cap.IRQControlCap
+   | cap.IRQHandlerCap irq \<Rightarrow> cap.IRQHandlerCap irq
+   | cap.Zombie r b n \<Rightarrow> cap.Zombie r b n
+   | cap.ArchObjectCap cap \<Rightarrow> cap.ArchObjectCap cap)"
+  apply (cases cap,
+         simp_all only: cap.simps update_cap_data_def is_ep_cap.simps if_False if_True
+                        is_ntfn_cap.simps is_cnode_cap.simps is_arch_cap_def word_size
+                        cap_ep_badge.simps badge_update_def o_def cap_rights_update_def
+                        simp_thms cap_rights.simps Let_def split_def
+                        the_cnode_cap_def fst_conv snd_conv fun_app_def the_arch_cap_def
+                        arch_update_cap_data_def
+                  cong: if_cong)
+  apply (auto simp: word_bits_def)
+  done
```
I'm not sure what's going on here. Is the copy in Ipc_A still needed?

mbrcknl replied:
This patch introduces copies in other arches as well.

lsf37 replied:
Well spotted, I’m not sure actually. This might have been an experiment on ARM which then got copied.  Will double check.

mbrcknl commented on proof/crefine/X64/Retype_C.thy:
```diff
@@ -3249,18 +3249,6 @@ lemma heap_list_update':
   "\<lbrakk> n = length v; length v \<le> 2 ^ word_bits \<rbrakk> \<Longrightarrow> heap_list (heap_update_list p v h) n p = v"
   by (simp add: heap_list_update addr_card_wb)
 
-lemma heap_update_field':
-  "\<lbrakk>field_ti TYPE('a :: packed_type) f = Some t; c_guard p;
-  export_uinfo t = export_uinfo (typ_info_t TYPE('b :: packed_type))\<rbrakk>
-  \<Longrightarrow> heap_update (Ptr &(p\<rightarrow>f) :: 'b ptr) v hp =
-  heap_update p (update_ti_t t (to_bytes_p v) (h_val hp p)) hp"
-  apply (erule field_ti_field_lookupE)
-  apply (subst packed_heap_super_field_update [unfolded typ_uinfo_t_def])
-     apply assumption+
-  apply (drule export_size_of [simplified typ_uinfo_t_def])
-  apply (simp add: update_ti_t_def)
-  done
-
```
Could this also be removed from `ARM` and `ARM_HYP`?

lsf37 replied:
Probably yes, need to check if it exists there.

mbrcknl commented on proof/crefine/X64/Syscall_C.thy:
```diff
@@ -358,20 +358,10 @@ lemma wordFromRights_mask_0:
   apply (simp add: wordFromRights_mask_0)
   done
 
-lemma loadWordUser_user_word_at:
-  "\<lbrace>\<lambda>s. \<forall>rv. user_word_at rv x s \<longrightarrow> Q rv s\<rbrace> loadWordUser x \<lbrace>Q\<rbrace>"
-  apply (simp add: loadWordUser_def user_word_at_def
-                   doMachineOp_def split_def)
-  apply wp
-  apply (clarsimp simp: pointerInUserData_def
-                        loadWord_def in_monad upto0_7_def
-                        is_aligned_mask)
-  done
-
```
Can we do the same thing in `ARM` and `ARM_HYP`?

mbrcknl commented on spec/abstract/ARM_HYP/ArchVSpace_A.thy:
```diff
@@ -537,7 +537,7 @@ where
 
           mapM (\<lambda>vreg. do
-                    val \<leftarrow> do_machine_op $ get_gic_vcpu_ctrl_lr (of_int vreg);
+                    val \<leftarrow> do_machine_op $ get_gic_vcpu_ctrl_lr (of_nat vreg);
```
`upt` always produced `nat`s, so I guess the type of `of_int` became a bit less permissive?

lsf37 replied:
Yes, of-int changed upstream and now actually wants an int. There were one or two more instances of this somewhere.

mbrcknl reviewed (approved):
Epic patch! My browser struggled a bit. Also looks like it includes some nice clean-up.

I've done a *very* superficial pass over the whole thing. (Found it easier to take it a commit at a time.) I obviously didn't take in the details of the Word_Lib reorganisation, but agree that it seems to have simplified some proofs.

lsf37 commented:
Addressed Matt's comments in a new commit and updated the testboard. I've only checked ARM_HYP so far, so there is chance that the other architectures are failing. We'll see.

The AWS test VM now has the Isabelle2021 contrib components and should not spend all its time on downloading them any more..

lsf37 commented:
There was still a slight tweak in Word_Lib necessary to avoid shadowing of the default word induct/cases rules.

If the testboard succeeds, is everyone happy for me to merge this or do you want to do more review?

mktnk3 commented on lib/Word_Lib/Ancient_Numeral.thy:
```diff
@@ -0,0 +1,235 @@
+(*
+ * Copyright 2020, Data61, CSIRO (ABN 41 687 119 230)
+ *
```
Do we need to worry about the license header here?

lsf37 replied:
It's mostly just material that has been moved around and/or slightly adjusted, so that header is accurate. But this is just imported from the AFP anyway, almost none of the changes in Word_Lib are ours.
